### ISSUE TITLE
Why does my model exists some negative weight edges?

### ISSUE THREAD

**Issue by 1021808202** [quantumblacklabs/causalnex#59]
Hi,
After learn the BN model,  I use the 'sm.egdes(data)' and then I find some negative weight edges in my model.
What's the meaning of those negative weight edges?

**Comment by SteveLerQB** [comment]
Hi @1021808202, thanks for your question. 

What the structure learning NOTEARS algorithm essentially does is essentially fitting a multilinear regression problem. Each weight represents an edge, and the algorithm tries to make sure that these edges which make up a graph is a non cyclic graph. 

The negative edge A->B with coefficient `w` means that there is a negative relationship between the A and B. In other words, one unit increase in A leads to a decrease of `w` amount in B. 

Hope this answers your question. :)

**Comment by SteveLerQB** [comment]
You can check this issue https://github.com/quantumblacklabs/causalnex/issues/9 for a toy example to play with

**Comment by 1021808202** [comment]
@SteveLerQB Thanks a lot.
However, I still wonder if those negative weight edges affect the result of BN inference?
In your example , the negative edge A->B with coefficient w((satisfy the DAG's requirements)), essentially, B->A, not A->B. when I do BN inference, if I give a value to B(child node), does A(parent node) can get a value?

**Comment by SteveLerQB** [comment]
Hi @1021808202, the negative weight edge does not affect the result of BN inference, in fact when fitting the BN model it does not take the weight coefficient into account, and you might expect something like: A->B, P(B>b | A<a) >  P(B>b | A>a)

A -> B wth a negative coefficient w does not mean B -> A. One example for this is increasing exercise level causes a decrease in body weight. Exercise causes the body weight, not the other way round, and it is a negative relationship. Hope this helps :)

**Comment by 1021808202** [comment]
@SteveLerQB Thanks for your careful explanation.

**Comment by 1021808202** [comment]
Hi @SteveLerQB , I'm still thinking about  the meaning of  the weight coefficient.

Is the egdes'  weight coefficient equivalent to the conditional probability ? For example, does  A->B with a weight coefficient 0.6 or -0.6 means the edges'  conditional probability is 0.6 ? As we all know, the BN use conditional probability to describe the edges' weight, so what's the difference between the egdes'  weight coefficient and the conditional probability?

As you said above，‘’The negative edge A->B with coefficient w means that there is a negative relationship between the A and B. In other words, one unit increase in A should lead to a decrease of w amount in B‘’  is the meaning of  the weight coefficient. And  when A happends, the probability of B happends is the meaning of  conditional probability. I'm really curious about the relationship.

Thanks a lot.

**Comment by SteveLerQB** [comment]
Hi @1021808202,

Thanks for your question again. Firstly, we **do not support Continuous Bayesian Network**, but I'd like to bring it here so you can get a clearer idea. Secondly, weight coefficient is a real number, it can range from negative infinity to positive infinity.

Imagine A->B and the **true relationship** is A = -0.6 * B + noise. If we fit the conditional probability for A, we will expect the conditional probability for A to be P(A| B=b) = N(-0.6*b, std(noise). In other words, **we expect the probability density of A=-0.6*b to be high**.

But since we are only supporting **Discrete Bayesian Network**, you can expect this relationship **P(B>b | A<a) > P(B>b | A>a)** in general (might not be the case if your causal graph is complicated).

Hope this helps! :)

**Comment by 1021808202** [comment]
Hi @SteveLerQB , 

Thanks again, however, I still want to show a example to get a better understanding under your kind help. 

Let's back to the docs' first example's Structure Learning part , the example see a edge with a weight coefficient below 0.8 as a weaker edge. I know the w_threshold need to be seen as a hyperparameter, however, I wonder if  a edge with a bigger weight coefficient(absolute value)  has a higher probability of  existing in the BN , or just use those edges with bigger weight coefficient to get the BN structure?

Also, I find that 'study time'(numeric) affects 'G1'(numeric) and 'failures'(numeric) affects 'absences'(numeric),which must have a positive egdes' weight coefficient. Assuming the first weight coefficient is 0.6 and the second weight coefficient is 0.3, can I 
expect 'study time'->'G1' is more important than 'failures' ->'absences' in the BN?

Ps: I've been preparing for my academic paper recently, so I have those questions. Your kindness help  likes a light in my dark, thank you very much. By the way ,could you please recommend something relevant to my question, such as some paper or books?

**Comment by SteveLerQB** [comment]
Hi @1021808202,

Thanks for your question again.

To get the best out of NOTEARS structure learning, I suggest you do the following:
1) Mean normalization (Minus data from your mean). Reason being is if your feature has a large mean, the weight coefficient will be small
2) Avoid running on categorical/Non-ordinal data  (This is still under WIP and we are hoping to find a way to do structure learning on these)

With the above then yes, a high weight coefficient suggests that there is relationship between the two variables, but also bear in mind that the NOTEARS structure learning can capture **linear relationships** well, it cannot capture **non-linear relationships**. Therefore I highly suggest to inspect the graph and see if the relationships make sense, and encode domain knowledge into the graph if necessary.

And there's the next question, how do I know if the structure of the Bayesian Network is a good structure? For this you have to run the probability fitting of the Bayesian Network, and inspect the classification score for each node. If the **classification score is bad then it is not a good structure**. **Iterate the process** by changing your structure until you find a good Bayesian Network structure which gives a good performance metric and explains the data well. I hope this helps. :)

**Comment by SteveLerQB** [comment]
Paper suggestions: 
1) https://dl.acm.org/doi/pdf/10.1145/3219819.3220104
2) https://papers.nips.cc/paper/8157-dags-with-no-tears-continuous-optimization-for-structure-learning.pdf
3) https://arxiv.org/abs/1909.13189

Book suggestion:
1) Probabilistic Graphical Models: Principles and Techniques

**Comment by 1021808202** [comment]
@SteveLerQB Thanks a lot.

**Comment by 1021808202** [comment]
Hi @SteveLerQB ,

Recently, I have got a new problem about the negative weight edges. I find some negative weight edges exist in my model when I use all nodes,  however, I think one edge (weight coefficient  -0.8) should be positive edge when  I check the relationship by observing the data . And then, I just use the negative weight edges' node( just 2 nodes ) to learn the relationship. The result is the edges change as positive edge ( weight coefficient  0.4). So I don't  know what happen and what should I do.

**Comment by oentaryorj** [comment]
Closing this issue for now, as most of the inquiries have been addressed as above. Regarding the last question of negative weights, this should be addressable by fine-tuning the hyperparameters such as 'W-Threshold' , 'beta' (for from_pandas_lasso) and/or running data preprocessing (e.g. scaling, discretizing, encoding etc.) until you get a graph that best matches your domain knowledge.
